File: Irrgarten_Java/src/irrgarten/Player.java
package irrgarten;

import java.util.ArrayList;

/**
 * Clase que almacena información de cada jugador
 * 
 * @author Arturo Olivares Martos
 * @author Joaquín Avilés de la Fuente
 */
public class Player extends LabyrinthCharacter {
    /**
     * Número máximo de armas que puede llevar un jugador.
     */
    private static final int MAX_WEAPONS=2;
    
    /**
     * Número máximo de escudos que puede llevar un jugador.
     */
    private static final int MAX_SHIELDS=3;
    
    /**
     * Número de unidades de salud inicial de los jugadores.
     */
    private static final int INTIAL_HEALTH=10;
   
    /**
     * Número de golpes que puede recibir un jugador antes de morir.
     */
    private static final int HITS2LOSE=3;

    
    
    
    /**
     * Número identificador del jugador.
     */
    private char number;
    /**
     * Número de golpes consecutivos que ha recibido el jugador.
     */
    private int consecutiveHits;
    
    /**
     * ArrayList de armas que lleva el jugador.
     */
    private ArrayList<Weapon> weapons;        // Se pone final porque no se asigna a otra cosa.
    /**
     * ArrayList de escudos que lleva el jugador.
     */
    private ArrayList<Shield> shields;        // Se pone final porque no se asigna a otra cosa.
    
    /**
     * Baraja de cartas de tipo Arma
     */
    private WeaponCardDeck weaponCardDeck;
    
    /**
     * Baraja de cartas de tipo Escudo
     */
    private ShieldCardDeck shieldCardDeck;
    
    /**
     * Constructor de la clase Player
     * @param number número identificador del jugador
     * @param intelligence inteligencia del jugador
     * @param strength fuerza del jugador
     */
    public Player (char number, float intelligence, float strength){
        super("Player "+number, intelligence, strength, INTIAL_HEALTH);
        this.number=number;
        this.consecutiveHits=0;

        // Hay que inicializar los ArrayList
        this.weapons= new ArrayList<>();
        this.shields= new ArrayList<>();
        
        // Hay que inicializar las barajas
        this.weaponCardDeck= new WeaponCardDeck();
        this.shieldCardDeck= new ShieldCardDeck();
    }
    
    /**
     * Consctructor de copia
     * @param other Objeto que copiar // TODO: preguntar a Lastra
     */
    public Player (Player other){
        super(other); // Se usa el constructor de copia de LabyrinthCharacter
        this.number=other.number;
        this.consecutiveHits=other.consecutiveHits;
        
        // Hay que inicializar los ArrayList, pero vacíos
        weapons= new ArrayList<>();
        shields= new ArrayList<>();
    }
    
    /**
     * Resurreción de un jugador, indicando los valores que debe tener
     * al resucitar
     */
    public void resurrect(){
        this.setHealth(INTIAL_HEALTH);
        resetHits();
        // Inicializamos los vectores como vacío
        this.weapons.clear();
        this.shields.clear();
    }

    /**
     * Informa sobre el número del jugador
     * @return Número del jugador
     */
    public char getNumber(){
        return this.number;
    }
    
    /**
     * Método que informa sobre la dirección en la que se va a mover el jugador
     * 
     * @param direction  Dirección en la que se quiere mover el jugador
     * @param validMoves  Lista de movimientos válidos
     * @return  Devuelve la dirección en la que se moverá el jugador (si es válida)
     */
    public Directions move(Directions direction, ArrayList<Directions> validMoves){
        int size = validMoves.size();
        boolean contained = validMoves.contains(direction);

        Directions toReturn;

        if ((size > 0) && !contained){ // El jugador se puede mover, pero no en la dirección deseada
            toReturn = validMoves.get(0);
        }
        else{
            toReturn = direction;
        }
        // Si size==0, se devuelve la dirección deseada y ya se comprobará en putPlayer2D que no se puede mover

        return toReturn;
    }
    
    /**
     * Calcula ataque total del jugador, teniendo en cuenta su fuerza y el poder
     * de sus armas
     * @return Devuelve la suma de su fuerza y el poder de las armas
     */
    @Override
    public float attack(){
        return (this.getStrength()+this.sumWeapons());
    }
    
    /**
     * Método que permite al jugador defenderse de un ataque.
     * Relega en manageHit(float) la gestión de la defensa.
     * @param receivedAttack Intensidad del ataque recibido
     * @return  Devuelve true si el jugador ha muerto y false en caso contrario.
     * @see #manageHit(float)
     */
    @Override
    public boolean defend(float receivedAttack){
        return this.manageHit(receivedAttack);
    }
    
    /**
     * Método que actualiza las armas, escudos y salud del jugador tras recibir una recompensa.     * 
     */
    public void receiveReward(){
        // Weapons rewards
        int wReward = Dice.weaponsReward();
        for (int i=0; i<wReward; i++)
            this.receiveWeapon(this.newWeapon());
        
        // Shields rewards
        int sReward = Dice.shieldsReward();
        for (int i=0; i<sReward; i++)
            this.receiveShield(this.newShield());

        // Health reward
        this.setHealth(this.getHealth()+Dice.healthReward());
    }
    
    /**
     * Método que devuelve una representación en forma de cadena de caracteres
     * del estado interno del jugador.
     * @return Representación en forma de cadena de caracteres del estado interno del jugador.
     */
    @Override // TODO: comprobar que está igual en ruby
    public String toString(){
        String toReturn=super.toString();
        toReturn+=", ch:"+this.consecutiveHits+", ";
        
        // Bucles para mostrar con un formato determinado el array de
        // armas y escudos del jugador
        String toWeapons="[";
        int tamWeapons=this.weapons.size();
        for(int i=0; i<tamWeapons-1; i++){
            toWeapons+=this.weapons.get(i).toString()+", ";
        }
        if (tamWeapons>0)
            toWeapons+=this.weapons.get(tamWeapons-1);
        toWeapons+="]";
        
        String toShields="[";
        int tamShields=this.shields.size();
        for(int i=0; i<tamShields-1; i++){
            toShields+=this.shields.get(i).toString()+", ";
        }
        if (tamShields>0)
            toShields+=this.shields.get(tamShields-1);
        toShields+="]";
        
        // Definimos el formato final para el toString
        toReturn+="w:" + toWeapons+", sh:"+toShields+" ]";
        
        return toReturn;
    }

    /**
     * Método que actualiza las armas del jugador tras recibir un arma.
     * En primer lugar, se eliminan las armas que han de ser descartadas.
     * Después, se añade el arma al jugador si cabe.
     * @param w Arma nueva a añadir
     */
    private void receiveWeapon(Weapon w){
        // Iteramos sobre las armas del jugador, viendo si tienen que ser descartadas
        for (int i=0; i<weapons.size(); i++){
            if (weapons.get(i).discard()){
                weapons.remove(i);
                i--;
            }
        }

        // Añadimos el escudo al jugador si cabe
        if (weapons.size() < MAX_WEAPONS)
            weapons.add(w);
    }
    
    /**
     * Método que actualiza los escudos del jugador tras recibir un escudo.
     * En primer lugar, se eliminan los escudos que han de ser descartados.
     * Después, se añade el escudo al jugador si cabe.
     * 
     * @param s  Escudo que recibe el jugador
     */
    private void receiveShield(Shield s){

        // Iteramos sobre los escudos del jugador, viendo si tienen que ser descartados
        for (int i=0; i<shields.size(); i++){
            if (shields.get(i).discard()){
                shields.remove(i);
                i--;
            }
        }

        // Añadimos el escudo al jugador si cabe
        if (shields.size() < MAX_SHIELDS)
            shields.add(s);
    }
    
    // TODO: mirar si lo que quiere son estos cambios en newWeapon y newShield
    /**
     * Genera una nueva instancia de la clase Weapon, con los parámetros
     * que indica el dado
     * @return Devuelve el arma creada de forma aleatoria con el dado
     */
    private Weapon newWeapon(){
        return this.weaponCardDeck.nextCard();
    }
    
    /**
     * Genera una nueva instancia de la clase Shield, con los parámetros
     * que indica el dado
     * @return Devuelve el escudo creado de forma aleatoria con el dado
     */
    private Shield newShield(){
        return this.shieldCardDeck.nextCard();
    }
    
    /**
     * Indica la suma total de los poderes de todas las armas que tiene
     * el jugador
     * @return Devuelve la suma de poderes de todas las armas
     */
    protected float sumWeapons(){
       int attack=0;
       for (int i=0; i<weapons.size(); i++){
           attack+=weapons.get(i).attack();
       } 
       return attack;
    }
    
    /**
     * Indica la suma total de protección de todos los escudos que tiene
     * el jugador
     * @return Devuelve la suma de protección de todas los escudos
     */
    protected float sumShields(){
       int protect=0;
       for (int i=0; i<shields.size(); i++){
           protect+=shields.get(i).protect();
       } 
       return protect;
    }
    
    /**
     * Calcula defensa total del jugador, teniendo en cuenta su inteligencia y 
     * la protección de sus escudos
     * @return Devuelve la suma de su inteligencia y la protección de los escudos
     */
    protected float defensiveEnergy(){
        return (this.getIntelligence()+this.sumShields());
    }
    
    /**
     * Método que gestiona el ataque recibido por el jugador.
     * 
     * @param receivedAttack Intensidad del ataque recibido
     * @return Devuelve true si el jugador ha muerto y false en caso contrario.
     */
    private boolean manageHit(float receivedAttack){

        // Se contabiliza el ataque recibido
        if (this.defensiveEnergy() < receivedAttack){
            // El jugador no se ha defendido
            this.gotWounded();
            this.incConsecutiveHits();
        }
        else{
            // El jugador se ha defendido
            this.resetHits();
        }

        // Se comprueba si el jugador ha perdido
        boolean lose = (this.consecutiveHits==Player.HITS2LOSE) || this.dead();

        if (lose)   // Si ha perdido se resetea el contador de golpes consecutivos
            resetHits();

        return lose;
    }

    /**
     * Reinicia el contador de impactos consecutivos a 0
     */
    private void resetHits(){
        this.consecutiveHits=0;
    }
    
    /**
     * Incrementa en una unidad el contador de impactos consecutivos
     */
    private void incConsecutiveHits(){
        this.consecutiveHits++;
    }
}

File: Irrgarten_Ruby/player.rb
# encoding: UTF-8

require_relative 'dice'
require_relative 'weapon'
require_relative 'shield'
require_relative 'labyrinth_character'


module Irrgarten

    # Clase que representa al jugador del juego.
    #
    #
    # @author Joaquin Avilés de la Fuente
    # @author Arturo Olivares Martos
    class Player < LabyrinthCharacter

        @@MAX_WEAPONS = 2       # Número máximo de armas que puede tener un jugador
        @@MAX_SHIELDS = 3       # Número máximo de escudos que puede tener un jugador
        @@INITIAL_HEALTH = 10   # Salud inicial de los jugadores
        @@HITS2LOSE = 3         # Número de golpes que puede recibir un jugador antes de morir

        # Constructor de la clase Player. Inicializa los atributos de la clase.
        # La posición inicial del jugador es inválida
        #
        # @param number [char] Número del jugador
        # @param intelligence [float] Inteligencia del jugador
        # @param strength [float] Fuerza del jugador
        def initialize(number, intelligence, strength)
            @number = number.to_s
            name = "Player #{@number}"
            super(name, intelligence, strength, @@INITIAL_HEALTH)

            @weapons = Array.new           # Array de armas
            @shields = Array.new           # Array de escudos

            @consecutive_hits = 0   # Número de golpes consecutivos recibidos. Al crearse está en 0
        end



        # Consultor de @number
        # @return [char] número del jugador
        attr_reader :number

        protected   # // TODO: Qué visibilidad tienen en Ruby??
        # Consultor de @weapons
        # @return [Array::Weapon] array de armas del jugador
        attr_reader :weapons

        # Consultor de @shields
        # @return [Array::Shield] array de escudos del jugador
        attr_reader :shields

        # Consultor de @consecutive_hits
        # @return [int] número de golpes consecutivos recibidos
        attr_reader :consecutive_hits
        public


        # Método que busca asemejarse a un constructor de copia.
        # Copia los atributos de un jugador a otro.
        #
        # @param other [Player] jugador al que se quiere copiar
        # @note Al terminar se copian los atributos del jugador **other** al jugador que llama al método
        def copy(other)
            super(other)
            @number = other.number
            @weapons = other.weapons
            @shields = other.shields
            @consecutive_hits = other.consecutive_hits
        end

        # Método que resucita al jugador, poniendo su salud a su valor inicial,
        # vaciando su lista de armas y escudos, y poniendo los golpes consecutivos a 0
        def resurrect
            @health = @@INITIAL_HEALTH
            @weapons.clear
            @shields.clear
            reset_hits()
        end


        # Comprueba si la dirección pasada hacia la que se pretende desplazar el personaje
        # es válida, devolviendola en caso de que lo sea o no se pueda mover hacia ninguna posición, es decir,
        # valid_moves esté vacío. Si no está en valid_moves y dicho array no está vacío, se devuelve la primera dirección
        # guardada en el array
        #
        # @param direction [Directions] dirección a la que se pretende desplazar el personaje
        # @param valid_moves [Array::Directions] lista de direcciones válidas a las que se puede mover el jugador
        #
        # @return [Directions] dirección a la que se quiere desplazar (tendremos que ver si es válida)
        def move(direction, valid_moves)
            size = valid_moves.length

            # El método del array es include?(<element>)
            contained = valid_moves.include?(direction)


            if ( (size>0) && (!contained) )
                return valid_moves[0] # Se puede también array.at(pos)
            else
                return direction
            end
        end

        # Calcula la suma de la fuerza del jugador y la suma de lo aportado por sus armas (sum_weapons).
        #
        # @return [float] la intensidad del ataque
        def attack
            return @strength + self.sum_weapons # Self no es necesario
        end

        # Este método delega su funcionalidad en el método manage_hit de Player
        # @see Player#manage_hit
        #
        # @param received_attack [float] ataque recibido
        #
        # @return [boolean] devuelve true si ha muerto, y false en caso contraio
        def defend(received_attack)
            return manage_hit(received_attack)
        end

        # Método que recompensa al jugador con armas, escudos y vida extra delegando
        # para ello en métodos de Dice, como weapons_reward, shields_reward y health_reward
        # @see Dice#weapons_reward
        # @see Dice#shields_reward
        # @see Dice#health_reward
        def receive_reward
            wReward=Dice.weapons_reward
            sReward=Dice.shields_reward

            # Reward de armas
            wReward.times do |i|
                receive_weapon(self.new_weapon)
            end

            # Reward de escudos
            sReward.times do |i|
                receive_shield(self.new_shield)
            end

            # Reward de vida
            @health += Dice.health_reward
        end

        # Método que genera una cadena de caracteres con la información del jugador
        #
        # @return [String] cadena de caracteres con la información del jugador
        def to_s
            # Guardamos la info de todas las armas en un string
            to_weapons="["
            tam_weapons=@weapons.size
            # Se incluye 0 y tam_weapons-1
            for i in 0..(tam_weapons-1) do
                to_weapons+=@weapons[i].to_s

                # NO hace la parte de delante si i == (tam_weapons-1)
                to_weapons += ", " unless i == (tam_weapons - 1)
            end
            to_weapons+="]"

            # Guardamos la info de todas las armas en un string
            to_shields="["
            tam_shields=@shields.size
            # Se incluye 0 y tam_shields-1
            for i in 0..(tam_shields-1) do
                to_shields+=@shields[i].to_s

                # NO hace la parte de delante si i == (tam_weapons-1)
                to_shields += ", " unless i == (tam_shields - 1)
            end
            to_shields+="]"

            return super + ", ch:#{@consecutive_hits}, w:"+to_weapons+", sh:"+to_shields + " ]"
        end

        private

        # Método que añade un nuevo arma (si es posible) al array de armas del jugador. Primero
        # intenta eliminar algún arma usando el método discard de Weapon y luego si el numero
        # de armas del jugador es menor estricto al numero de armas máximas a llevar, se añade
        # el nuevo arma.
        # @see Weapon#discard
        #
        # @param w [Weapon] arma a intentar añadir al jugador
        def receive_weapon(w)

            # En primer lugar, eliminamos las armas debidas
            @weapons.delete_if do |wi|
                wi.discard
            end

            # Si caben, se añade el arma
            if(@weapons.length < @@MAX_WEAPONS)
                @weapons.push(w) # funciona también @weapons<<(wi)
            end
        end

        # Método que añade un nuevo escudo (si es posible) al array de escudos del jugador. Primero
        # intenta eliminar algún escudo usando el método discard de Shield y luego si el número
        # de escudos del jugador es menor estricto al numero de escudos máximos a llevar, se añade
        # el nuevo escudo.
        # @see Shield#discard
        #
        # @param s [Shield] escudo a intentar añadir al jugador
        def receive_shield(s)
            @shields.delete_if do |si|
                si.discard
            end

            if(@shields.length < @@MAX_SHIELDS)
                @shields.push(s)
            end
        end

        # Método que genera una nueva arma
        #
        # @return [Weapom] devuelve un nuevo objeto de tipo Weapon. NO lo añade a la lista de armas
        def new_weapon
            return Weapon.new(Dice.weapon_power, Dice.uses_left)
        end

        # Método que genera un nuevo escudo
        #
        # @return [Shield] devuelve un nuevo objeto de tipo Shield. NO lo añade a la lista de escudos
        def new_shield
            return Shield.new(Dice.shield_power, Dice.uses_left)
        end

        # // TODO: Protected en ruby es igual que en Java??

        protected
        # Devuelve la suma del resultado de llamar al método attack de todas sus armas.
        #
        # @return [float] la apotación de las armas al ataque
        def sum_weapons
            sum = 0
            @weapons.each do |w|
                sum += w.attack
            end
            return sum
        end

        # Devuelve la suma del resultado de llamar al método protect de todos sus escudos.
        #
        # @return [float] la apotación de los escudos a la defensa
        def sum_shields
            sum = 0
            @shields.each do |s|
                sum += s.protect
            end
            return sum
        end

        # Calcula la suma de la inteligencia con el aporte de los escudos (sum_shields).
        #
        # @return [float] la energía defensiva
        def defensive_energy
            return @intelligence + sum_shields
        end
        public

        # Método que gestiona la defensa de un ataque. En el caso de que el jugador
        # haya recibido un número máximo de golpes consecutivos, determinado por
        # @@HITS2LOSE, se considerará que ha perdido, al igual que si muere
        #
        # @param received_attack [float] ataque a defender
        #
        # @return [boolean] devuelve true si ha muerto o ha alcanzado el máximo número
        # de ataques, y false en caso contrario
        def manage_hit(received_attack)
            defense=self.defensive_energy
            if(defense<received_attack)
                self.got_wounded # recibe daño
                self.inc_consecutive_hits
            else
                self.reset_hits
            end

            if ( (@consecutive_hits==@@HITS2LOSE) || self.dead )
                self.reset_hits
                lose=true
            else
                lose=false
            end
            return lose
        end

        # Fija el valor del contador de impactos consecutivos a cero.
        def reset_hits
            @consecutive_hits = 0
        end

        # Incrementa en una unidad el contador de impactos consecutivos.
        def inc_consecutive_hits
            @consecutive_hits += 1
        end

    end # class
end # module
